[FEATURE] Data source for databricks_service_principal

Hi there,
I tried to assign an existing service principal to existing group and I get the following error:
```sh
The provider databrickslabs/databricks does not support data source "databricks_service_principal".

Did you intend to use the managed resource type "databricks_service_principal"? If so, declare this using a "resource" block instead of a "data" block.
```

When I use the resource block I am not able to modify this principal later on , it gives the following error:
```sh
Error: cannot update service principal: Request is unparsable, syntactically incorrect, or violates schema.
```

Another issue is that it creates another brand new service principal with the same display name when try specifying it in another terraform configuration, i.e. **databricks_obo_token** resource configuration which resides in another directory within our project.

Is there a reason why I cannot refer to existing service principal? Is there a feature request for the same?

### Configuration

#### main.tf
```hcl
data "databricks_user" "users" {
  for_each  = toset(local.users)
  user_name = each.value
}

data "databricks_group" "groups" {
  for_each     = var.groups
  display_name = each.key
}

data "databricks_service_principal" "sp" {
  for_each                   = var.global_service_principals
  display_name               = each.value.display_name
}


locals {
  users = flatten([
  for group in keys(var.groups) : [
  for user in var.groups[group].users : [user]
  ]
  ])

  service_principals = flatten([
  for group in keys(var.groups) : [
  for sp in var.groups[group].service_principals : [var.global_service_principals[sp]]
  ]
  ])

  user_group_association = flatten([
  for group in keys(var.groups) : [
  for user in var.groups[group].users :
  {
    user_name : user
    group : group
  }
  ]
  ])
  sp_group_association   = flatten([
  for group in keys(var.groups) : [
  for service_principal in var.groups[group].service_principals :
  {
    service_principal : service_principal
    group : group
  }
  ]
  ])
}

resource "databricks_group_member" "member" {
  for_each   = {
  for index, association in local.user_group_association :
  "${association["user_name"]} (${association["group"]})" => association
  }
  group_id   = data.databricks_group.groups[each.value["group"]].id
  member_id  = data.databricks_user.users[each.value["user_name"]].id
  depends_on = [data.databricks_user.users, data.databricks_group.groups]
}
resource "databricks_group_member" "member_sp" {
  for_each   = {
  for index, association in local.sp_group_association :
  "${association["service_principal"]} (${association["group"]})" => association
  }
  group_id   = data.databricks_group.groups[each.value["group"]].id
  member_id  = data.databricks_service_principal.sp[each.value["service_principal"]].id
  depends_on = [data.databricks_service_principal.sp, data.databricks_group.groups]
}
```

#### variables.tf
```hcl
variable "groups" {
  type = map(object({
    users : list(string)
    service_principals : list(string)
  }))
}

variable "global_service_principals" {
  type = map(object({
    display_name               = string
    application_id             = optional(string)
    allow_cluster_create       = optional(bool)
    allow_instance_pool_create = optional(bool)
    databricks_sql_access      = optional(bool)
    workspace_access           = optional(bool)
    active                     = optional(bool)
  }))
}

```

#### versions.tf
```hcl
terraform {
  experiments = [module_variable_optional_attrs]
  required_providers {
    databricks = {
      source  = "databrickslabs/databricks"
      version = "0.4.5"
    }
  }
}
```


### Expected Behavior
Terraform should have found the existing service principal and assign it to a group in Databricks

### Actual Behavior
Creates new service principal besides having same-name one already existing in Databricks. Also, fails to update the newly created one later on. 

### Steps to Reproduce
Please run the following command:

```sh
terraform plan -var='groups={"Analytics":{ "users": [],"service_principals": ["insights-service-principal"]}}' -var='global_service_principals={"insights-service-principal": {display_name: "insights-service-principal"}}'
```

The error should be output right after executing it.
 
### Terraform and provider versions

0.4.5 (latest)

### Debug Output

```sh
TF_LOG=DEBUG terraform plan -var='groups={"Analytics":{ "users": [],"service_principals": ["insights-service-principal"]}}' -var='global_service_principals={"insights-service-principal": {display_name: "insights-service-principal"}}'

2022-01-20T10:46:57.965-0800 [DEBUG] Adding temp file log sink: /var/folders/tp/stprhcpd0w77_7wv_xy_m6m80000gp/T/terraform-log685109055
2022-01-20T10:46:57.966-0800 [INFO]  Terraform version: 0.15.5
2022-01-20T10:46:57.966-0800 [INFO]  Go runtime version: go1.16.4
2022-01-20T10:46:57.966-0800 [INFO]  CLI args: []string{"/usr/local/bin/terraform", "plan", "-var=groups={\"Analytics\":{ \"users\": [],\"service_principals\": [\"insights-service-principal\"]}}", "-var=global_service_principals={\"insights-service-principal\": {display_name: \"insights-service-principal\"}}"}
2022-01-20T10:46:57.966-0800 [DEBUG] Attempting to open CLI config file: /Users/martino.nikolovski/.terraformrc
2022-01-20T10:46:57.966-0800 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2022-01-20T10:46:57.966-0800 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2022-01-20T10:46:57.966-0800 [DEBUG] ignoring non-existing provider search directory /Users/martino.nikolovski/.terraform.d/plugins
2022-01-20T10:46:57.966-0800 [DEBUG] ignoring non-existing provider search directory /Users/martino.nikolovski/Library/Application Support/io.terraform/plugins
2022-01-20T10:46:57.966-0800 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2022-01-20T10:46:57.966-0800 [INFO]  CLI command args: []string{"plan", "-var=groups={\"Analytics\":{ \"users\": [],\"service_principals\": [\"insights-service-principal\"]}}", "-var=global_service_principals={\"insights-service-principal\": {display_name: \"insights-service-principal\"}}"}
2022-01-20T10:46:57.969-0800 [DEBUG] New state was assigned lineage "6ec9839d-024e-a063-c2dc-724b59fceef9"
2022-01-20T10:46:58.049-0800 [DEBUG] checking for provisioner in "."
2022-01-20T10:46:58.060-0800 [DEBUG] checking for provisioner in "/usr/local/bin"
2022-01-20T10:46:58.061-0800 [INFO]  Failed to read plugin lock file .terraform/plugins/darwin_amd64/lock.json: open .terraform/plugins/darwin_amd64/lock.json: no such file or directory
2022-01-20T10:46:58.061-0800 [INFO]  backend/local: starting Plan operation
2022-01-20T10:46:58.064-0800 [DEBUG] created provider logger: level=debug
2022-01-20T10:46:58.064-0800 [INFO]  provider: configuring client automatic mTLS
2022-01-20T10:46:58.103-0800 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5 args=[.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5]
2022-01-20T10:46:58.107-0800 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5 pid=34762
2022-01-20T10:46:58.107-0800 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5
2022-01-20T10:46:58.127-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: Databricks Terraform Provider (experimental)
2022-01-20T10:46:58.127-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: 
2022-01-20T10:46:58.127-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: Version 0.4.5
2022-01-20T10:46:58.127-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: 
2022-01-20T10:46:58.127-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: https://registry.terraform.io/providers/databrickslabs/databricks/latest/docs
2022-01-20T10:46:58.127-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: 
2022-01-20T10:46:58.131-0800 [INFO]  provider.terraform-provider-databricks_v0.4.5: configuring server automatic mTLS: timestamp=2022-01-20T10:46:58.130-0800
2022-01-20T10:46:58.177-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: plugin address: network=unix address=/var/folders/tp/stprhcpd0w77_7wv_xy_m6m80000gp/T/plugin380749225 timestamp=2022-01-20T10:46:58.177-0800
2022-01-20T10:46:58.177-0800 [DEBUG] provider: using plugin: version=5
2022-01-20T10:46:58.246-0800 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-01-20T10:46:58.250-0800 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5 pid=34762
2022-01-20T10:46:58.251-0800 [DEBUG] provider: plugin exited
2022-01-20T10:46:58.252-0800 [INFO]  terraform: building graph: GraphTypeValidate
2022-01-20T10:46:58.253-0800 [DEBUG] ProviderTransformer: "data.databricks_user.users" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/databrickslabs/databricks"]
2022-01-20T10:46:58.253-0800 [DEBUG] ProviderTransformer: "data.databricks_group.groups" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/databrickslabs/databricks"]
2022-01-20T10:46:58.253-0800 [DEBUG] ProviderTransformer: "data.databricks_service_principal.sp" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/databrickslabs/databricks"]
2022-01-20T10:46:58.254-0800 [DEBUG] ProviderTransformer: "databricks_group_member.member" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/databrickslabs/databricks"]
2022-01-20T10:46:58.254-0800 [DEBUG] ProviderTransformer: "databricks_group_member.member_sp" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/databrickslabs/databricks"]
2022-01-20T10:46:58.254-0800 [ERROR] AttachSchemaTransformer: No resource schema available for data.databricks_service_principal.sp
2022-01-20T10:46:58.254-0800 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/databrickslabs/databricks\"]" references: []
2022-01-20T10:46:58.254-0800 [INFO]  ReferenceTransformer: reference not found: "each.value"
2022-01-20T10:46:58.254-0800 [DEBUG] ReferenceTransformer: "data.databricks_user.users" references: [local.users (expand)]
2022-01-20T10:46:58.254-0800 [INFO]  ReferenceTransformer: reference not found: "each.key"
2022-01-20T10:46:58.254-0800 [DEBUG] ReferenceTransformer: "data.databricks_group.groups" references: [var.groups]
2022-01-20T10:46:58.254-0800 [WARN]  no schema is attached to data.databricks_service_principal.sp, so config references cannot be detected
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "data.databricks_service_principal.sp" references: [var.global_service_principals var.global_service_principals]
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "local.service_principals (expand)" references: [var.groups var.groups var.global_service_principals]
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "local.user_group_association (expand)" references: [var.groups var.groups]
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "local.sp_group_association (expand)" references: [var.groups var.groups]
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "local.users (expand)" references: [var.groups var.groups]
2022-01-20T10:46:58.255-0800 [INFO]  ReferenceTransformer: reference not found: "each.value"
2022-01-20T10:46:58.255-0800 [INFO]  ReferenceTransformer: reference not found: "each.value"
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "databricks_group_member.member" references: [data.databricks_user.users data.databricks_group.groups local.user_group_association (expand) data.databricks_user.users data.databricks_group.groups]
2022-01-20T10:46:58.255-0800 [INFO]  ReferenceTransformer: reference not found: "each.value"
2022-01-20T10:46:58.255-0800 [INFO]  ReferenceTransformer: reference not found: "each.value"
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "databricks_group_member.member_sp" references: [data.databricks_service_principal.sp data.databricks_group.groups local.sp_group_association (expand) data.databricks_service_principal.sp data.databricks_group.groups]
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "var.groups" references: []
2022-01-20T10:46:58.255-0800 [DEBUG] ReferenceTransformer: "var.global_service_principals" references: []
2022-01-20T10:46:58.255-0800 [WARN]  no schema is attached to data.databricks_service_principal.sp, so config references cannot be detected
2022-01-20T10:46:58.256-0800 [DEBUG] Starting graph walk: walkValidate
2022-01-20T10:46:58.257-0800 [DEBUG] created provider logger: level=debug
2022-01-20T10:46:58.257-0800 [INFO]  provider: configuring client automatic mTLS
2022-01-20T10:46:58.304-0800 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5 args=[.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5]
2022-01-20T10:46:58.307-0800 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5 pid=34763
2022-01-20T10:46:58.307-0800 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5
2022-01-20T10:46:58.324-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: Databricks Terraform Provider (experimental)
2022-01-20T10:46:58.324-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: 
2022-01-20T10:46:58.324-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: Version 0.4.5
2022-01-20T10:46:58.324-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: 
2022-01-20T10:46:58.324-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: https://registry.terraform.io/providers/databrickslabs/databricks/latest/docs
2022-01-20T10:46:58.324-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: 
2022-01-20T10:46:58.328-0800 [INFO]  provider.terraform-provider-databricks_v0.4.5: configuring server automatic mTLS: timestamp=2022-01-20T10:46:58.328-0800
2022-01-20T10:46:58.378-0800 [DEBUG] provider.terraform-provider-databricks_v0.4.5: plugin address: address=/var/folders/tp/stprhcpd0w77_7wv_xy_m6m80000gp/T/plugin070996662 network=unix timestamp=2022-01-20T10:46:58.378-0800
2022-01-20T10:46:58.379-0800 [DEBUG] provider: using plugin: version=5
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│ 
│   on versions.tf line 2, in terraform:
│    2:   experiments = [module_variable_optional_attrs]
│ 
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│ 
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
╵
╷
│ Error: Invalid data source
│ 
│   on main.tf line 10, in data "databricks_service_principal" "sp":
│   10: data "databricks_service_principal" "sp" {
│ 
│ The provider databrickslabs/databricks does not support data source "databricks_service_principal".
│ 
│ Did you intend to use the managed resource type "databricks_service_principal"? If so, declare this using a "resource" block instead of a "data" block.
╵
2022-01-20T10:46:58.455-0800 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-01-20T10:46:58.457-0800 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/databrickslabs/databricks/0.4.5/darwin_amd64/terraform-provider-databricks_v0.4.5 pid=34763
2022-01-20T10:46:58.457-0800 [DEBUG] provider: plugin exited
```

### Important Factoids
N/A

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,12 @@
 
 ## 0.4.7
 * Added optional `force` argument to `databricks_group` resource to ignore `cannot create group: Group with name X already exists.` errors and implicitly import the specific group into Terraform state, enforcing entitlements defined in the instance of resource ([#1066](https://github.com/databrickslabs/terraform-provider-databricks/pull/1066)).
+* Fixed `databricks_service_principal` `display_name` update ([#1065](https://github.com/databrickslabs/terraform-provider-databricks/issues/1065)).
+* Added documentation for Unity Catalog resources.
+
+Updated dependency versions:
+
+* Bump gopkg.in/ini.v1 from 1.66.2 to 1.66.3
 
 ## 0.4.6
 

--- a/common/http.go
+++ b/common/http.go
@@ -43,6 +43,7 @@ type APIErrorBody struct {
 	// for RFC 7644 Section 3.7.3 https://tools.ietf.org/html/rfc7644#section-3.7.3
 	ScimDetail string `json:"detail,omitempty"`
 	ScimStatus string `json:"status,omitempty"`
+	ScimType   string `json:"scimType,omitempty"`
 	API12Error string `json:"error,omitempty"`
 }
 
@@ -208,6 +209,9 @@ func (c *DatabricksClient) parseError(resp *http.Response) APIError {
 		} else {
 			errorBody.Message = errorBody.ScimDetail
 		}
+		// add more context from SCIM responses
+		errorBody.Message = fmt.Sprintf("%s %s", errorBody.ScimType, errorBody.Message)
+		errorBody.Message = strings.Trim(errorBody.Message, " ")
 		errorBody.ErrorCode = fmt.Sprintf("SCIM_%s", errorBody.ScimStatus)
 	}
 	if resp.StatusCode == 403 {

--- a/scim/resource_service_principal.go
+++ b/scim/resource_service_principal.go
@@ -69,7 +69,7 @@ func ResourceServicePrincipal() *schema.Resource {
 			m["active"].Default = true
 			return m
 		})
-	spFromData := func(d *schema.ResourceData) (user User, err error) {
+	spFromData := func(d *schema.ResourceData) User {
 		var u entity
 		common.DataToStructPointer(d, servicePrincipalSchema, &u)
 		return User{
@@ -77,7 +77,7 @@ func ResourceServicePrincipal() *schema.Resource {
 			DisplayName:   u.DisplayName,
 			Active:        u.Active,
 			Entitlements:  readEntitlementsFromData(d),
-		}, nil
+		}
 	}
 	return common.Resource{
 		Schema: servicePrincipalSchema,
@@ -95,10 +95,7 @@ func ResourceServicePrincipal() *schema.Resource {
 			return nil
 		},
 		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
-			sp, err := spFromData(d)
-			if err != nil {
-				return err
-			}
+			sp := spFromData(d)
 			if c.IsAws() && sp.ApplicationID != "" {
 				return fmt.Errorf("application_id is not allowed for service principals in Databricks on AWS")
 			}
@@ -121,11 +118,11 @@ func ResourceServicePrincipal() *schema.Resource {
 			return sp.Entitlements.readIntoData(d)
 		},
 		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
-			sp, err := spFromData(d)
-			if err != nil {
-				return err
-			}
-			return NewServicePrincipalsAPI(ctx, c).Update(d.Id(), sp)
+			return NewServicePrincipalsAPI(ctx, c).Update(d.Id(), User{
+				DisplayName:  d.Get("display_name").(string),
+				Active:       d.Get("active").(bool),
+				Entitlements: readEntitlementsFromData(d),
+			})
 		},
 		Delete: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
 			return NewServicePrincipalsAPI(ctx, c).Delete(d.Id())

